README
*******************
Note: Yaku-NS is a 10 years old project. It is here for personal "historical" reasons, as this is one of the first non trivial C projects I wrote.
*******************

+==================================================+
|                     WARNING                      |
|  Yaku-NS is in the beta stage of development     |
| possibly unstable, insecure, sperimental code.   |
|               USE IT AT YOUR RISK.               |
+==================================================+

-----------
ADVERTISING
-----------

  Yaku-ns is GPLed, but you can get a BSD-like license for your
  own company usage (for example to develop a closed-source fork
  and put yaku-ns in some router, appliance, ...) for 1000 EUR.
  The licence can't be sold to others, and applies to the current
  CVS source code, that is, you can't use code that I put
  in the CVS _after_ you already got the license.
  There aren't time limits. After you got the license you can
  modify the code and sell it in binary form, inside a ROM,
  where you want, forever.

  For more information please write to antirez@invece.org

--------
OVERVIEW
--------

  Yaku-NS is a DNS server that implements a subset of the
  DNS protocol.

  For a full featured DNS server check the BIND
  For a very secure DNS server (not complete, with a silly license) check DJBDNS
  For an alternative (almost complete) DNS server implementation check DENTS
  For a new upcoming GPL DNS implementation check MARADNS

  The goal is to provide an easy to configure GPLed alternative for:

	* A DNS server for myself. I'm unable to configure bind
	* A primary/secondary DNS server
	* Forward only DNS server
	* Dial-up systems
	* Embedded systems

----------------------------------
HOW TO SETUP YAKU IN YOUR UNIX BOX
----------------------------------

  Yaku-NS runs without root privileges in a chroot jail.
  We provided this features for your security, please enable
  this features!

COMMAND LINE OPTIONS
~~~~~~~~~~~~~~~~~~~~

  usage: yaku-ns [-p <port>] [-f <port>] [-C <max>] [-F <max>]
                 [-T <forward_timeout>] [-c <config_file>]
                 [-l <logfile>] [-r <chroot jail>] [-u owner]
                 [-b <addr>] [-xdDhV]

-p <port>	Use the UDP port <port> for the DNS service.

-P <port>	Use the TCP port <port> for the DNS under TCP service.
		The only two requests accepted under TCP are IN/AXFR and IN/SOA.

-f <port>	Query the external DNS servers to the port <port>.

-C <size>	Use a cache of size <size>, 0 is no-cache.

-F <max>	Use a forwarded request queue of size <size>, 0 turn off
		the forwarding.

-T <timeout>	Forwarded requets expire in <timeout> seconds.

-c <file>	Use the config file <file>: WARNING you must specify
		the absolute path here, since the server chdir() to "/".
		Also note that this path is relative to the chroot jail.
		Use "-" as filename to provide the configuration from the
		standard input.

-l <file>	Use the <file> for yaku-ns logs. Remember to use an
		absolute path, relative to the chroot (if used).

-r <directory>	Chroot to <directory>.

-u <username>	Run as <username> (default is nobody, not so secure).

-b <address>	Bind only the interface with IP address <address>.

-x		Enable the TCP services.

-d		Demonize.

-h		Show a little help.

-V		Verbosity level: -V: low level, -VV medium level, -VVV
		high level of verbosity. -VVV is raccomanded for new users.

  See how to create a configuration file using the self-commented
  config file at Documentation/yaku-ns.conf.example

  Follow this steps to install Yaku-NS in a unix-like system:

COMPILING YAKU-NS
~~~~~~~~~~~~~~~~~

  $ vi tunable.h (optional)
  $ make

INSTALLING ENS
~~~~~~~~~~~~~~

  Create a system user "yaku" in the "yaku" group.

  # mkdir /usr/local/yaku-ns
  # chown yaku:yaku /usr/local/yaku-ns
  # chmod 700 /usr/local/yaku-ns
  # cp yaku-ns /usr/local/yaku-ns
  # cp Documentation/yaku-ns.conf.example /usr/local/yaku-ns/yaku-ns.conf
  # chown root:root /usr/local/yaku-ns/*
  # chmod 755 /usr/local/yaku-ns/yaku-ns
  # chmod 644 /usr/local/yaku-ns/yaku-ns.conf
  # touch /usr/local/yaku-ns/yaku-ns.log
  # chown yaku:yaku /usr/local/yaku-ns/yaku-ns.log
  # chmod 644 /usr/local/yaku-ns/yaku-ns.log
  # cd /usr/local/yaku-ns
  # ls -l

  The output of the ls -l command should be like the following:

  total 168
  -rwxr-xr-x    1 root     root       153306 Dec 26 14:48 yaku-ns
  -rw-r--r--    1 root     root        10107 Dec 26 14:49 yaku-ns.conf
  -rw-r--r--    1 yaku     yaku            0 Dec 26 14:53 yaku-ns.log

CONFIGURING YAKU-NS
~~~~~~~~~~~~~~~~~~~

  Edit the example self-commented configuration file
  /usr/local/yaku-ns/yaku-ns.conf and create your configuration.

RUNNING YAKU-NS
~~~~~~~~~~~~~~~

  To run Yaku-NS inside a chroot jail (raccomanded!)
  with TCP services enabled just use:

  # /usr/local/yaku-ns/yaku-ns -xd -u yaku -r /usr/local/yaku-ns/ -c /yaku-ns.conf

  Note that the configuration file is specified using the absolute
  path, relative to the chroot.

SIGNALS
~~~~~~~

  Yaku-NS performs different action if some signal is received:

  SIGHUP:	Reload the configuration (use it when your records was updated)
  SIGUSR1:	Force yaku-ns to log some debugging information
  SIGUSR2:	Switch the forwarding ON/OFF

-------
GETZONE
-------

  Getzone is a trivial utility that performs an AXFR request
  over TCP to some primary DNS server and produces as output
  the zone in a format compatible with the Yaku-NS configuration file.

  You can use it to create a rudimental secondary DNS (slave)
  but a better solution will be on the road ASAP (It wasn't ASAP
  since I use only external secondary nameservers, so yaku-ns
  is my primary DNS server).

  Usage: getzone -z <zone> -s <server IP address> [ -p <port> ]

    -z		The zone to download
    -s		The IP address of the master server
    -p		Specify a different destination port, the default is port 53

  Example:

  $ getzone -z test.org -s 1.2.3.4 > salve-db.test.org

  Than you need to send a SIGHUP to Yaku-NS to force a local RRs reload.
  Obviously you need to add an include keyword in the yaku-ns.conf
  to include the zone. See the example configuration file for more
  information.

Have fun,
antirez


CHANGES
CHANGES file

Sep 16 2002		CNAME implemented
Sep 14 2002		Fixed a wildcard RR response building bug.
Sep 14 2002             Fixed a memory leak introduced with the new forwarding
                        ACL control.
Sep 12 2002             Forwarding fixed, now after the forward_next timeout
                        the query is sent to the next nameserver and so on.
                        Thanks to "Robert Allmeroth" <Robert.Allmeroth@epygi.de>
Sep 09 2002		Now stuff that should never be < 0 are unsigned.
Sep 09 2002		Major change, merged with the new hashtable library
			version.
Sep 08 2002		Implemented wildcard Resource Records, see Docs.
Sep 06 2002		Fixed a problem in name_decode(). If the name format
			was bad, but the pointer-to-pointer where to
			store the name was set to NULL (used to only know
			the length of the encoded name by get_min_ttl())
			the function deferenced a NULL pointer, writing
			0 on it. Not exploitable to break in but a bad DoS.
Sep 05 2002		now chdir() before to chroot().
Sep 04 2002		call abort() in the SIG_SEGV sighandler to dump the core
Sep 04 2002		use setrlimit(2) to enable core dumping
Sep 04 2002		Modified Copyright notices and email
Sep 04 2002		Fix for case insensitive hashtable key creation
Sep 03 2002		keyword 'forwarder' changed in 'nameserver'
Sep 03 2002		RRs qtype to symbol table updated
Sep 03 2002		Now names are matched case insensitive
Aug 30 2002		Fixed SIGCHLD handling problem.
Aug 30 2002		Fixed autoptr, now generates a PTR record for all the
			names for the IP address, not only for the first.
Aug 30 2002		New 'fwd' ACL chain used for forwarding access control
Feb  8 2001		Different algorithm for cache expiration check
Feb  7 2001		Now uses strlcpy and strlcat
Feb  6 2001		Zone tranfer code updated
Feb  5 2001		Fork from ENS:
			o Now uses hash table instead of lists
			o #ifdef away
			o config.c update
???  ? ????		o too much stuff, I don't remember


misc.c
/* misc.c
 * miscellaneous code
 *
 * Copyright (C) 2000,2001,2002 by Salvatore Sanfilippo
 * <antirez@invece.org>
 *
 * This code is under the GPL license
 * See the COPYING file for more information
 */

/* ens.h must be included before all other includes */
#include "ens.h"
#include "aht.h"

#ifdef __unix__
#include <sys/time.h>
#include <unistd.h>
#endif /* __unix__ */
#include <time.h>
#include <string.h>
#include <stdlib.h>

/* not exported functions */

/* exported functions */
u_int32_t get_rand32(void);
time_t get_sec(void);
char *qtype_to_str(unsigned short qtype);
char *qclass_to_str(unsigned short qclass);
int line_splitter(char *buffer, char *commandargs[], int argmax);
void dump_state(void);
u_int16_t get_rand_id(void);

/* --------------------------------------------------------------------------- */

time_t get_sec(void)
{
	return time(NULL);
}

char *qtype_to_str(unsigned short qtype)
{
	int i;
	static struct _QTYPE {
		unsigned short code;
		char *str;
	} qtype_str[] = {
		{ 1, "A" },
		{ 2, "NS" },
		{ 3, "MD" },
		{ 4, "MF" },
		{ 5, "CNAME" },
		{ 6, "SOA" },
		{ 7, "MB" },
		{ 8, "MG" },
		{ 9, "MR" },
		{ 10, "NULL" },
		{ 11, "WKS" },
		{ 12, "PTR" },
		{ 13, "HINFO" },
		{ 14, "MINFO" },
		{ 15, "MX" },
		{ 16, "TXT" },
		{ 17, "RP" },
		{ 18, "AFSDB" },
		{ 19, "X25" },
		{ 20, "ISDN" },
		{ 21, "RT" },
		{ 22, "NSAP" },
		{ 23, "NSAP_PTR"},
		{ 24, "SIG"},
		{ 25, "KEY"},
		{ 26, "PX"},
		{ 27, "GPOS"},
		{ 28, "AAAA"},
		{ 29, "LOC"},
		{ 30, "NXT"},
		{ 31, "EID"},
		{ 32, "NIMLOC"},
		{ 33, "SRV"},
		{ 34, "ATMA"},
		{ 35, "NAPTR"},
		{ 36, "KX"},
		{ 37, "CERT"},
		{ 38, "A6"},
		{ 39, "DNAME"},
		{ 40, "SINK"},
		{ 41, "OPT"},
/* non standard */
		{ 100, "UINFO"},
		{ 101, "UID"},
		{ 102, "GID"},
		{ 103, "UNSPEC"},
/* Query type values which do not appear in resource records */
		{ 249, "TKEY"},
		{ 250, "TSIG"},
		{ 251, "IXFR"},
		{ 252, "AXFR" },
		{ 253, "MAILB" },
		{ 254, "MAILA" },
		{ 255, "*" },
		{ 256, "bind-ZXFR"},
		{ 0, NULL} /* NUL TERM */
	};
	for (i = 0; qtype_str[i].code; i++) {
		if (qtype_str[i].code == qtype)
			return qtype_str[i].str;
	}
	return "UNKNOWN";
}

char *qclass_to_str(unsigned short qclass)
{
	int i;
	static struct _QCLASS {
		unsigned short code;
		char *str;
	} qclass_str[] = {
		{ 1, "IN" },
		{ 2, "CS" },
		{ 3, "CH" },
		{ 4, "HS" },
		{ 255, "*" },
		{ 0, NULL} /* NUL TERM */
	};

	for (i = 0; qclass_str[i].code; i++) {
		if (qclass_str[i].code == qclass)
			return qclass_str[i].str;
	}
	return "UNKNOWN";
}

#define skip_spacetab() while(*p == ' ' || *p == '\t') p++
int line_splitter(char *buffer, char *commandargs[], int argmax)
{
	char *p = buffer, *d;
	char tmp[1024];
	int size;
	int argindex = 0;

	/* if buffer is a NULL pointer free commandargs memory */
	if (buffer == NULL) {
		for (; commandargs[argindex] != NULL; argindex++)
			free(commandargs[argindex]);
		return argindex;
	}

	/* otherwise parse the command line */
	while(*p != '\0') {
		size = 0;
		d = tmp;
		skip_spacetab();

		while(*p != ' ' && *p != '\t') {
			if (*p == '\0' || *p == '\n' ||
			    *p == '\r' || size >= 1023)
				break;
			*d++ = *p++;
			size++;
		}

		if (size != 0) {
			commandargs[argindex] = malloc(size+1);
			if (commandargs[argindex] == NULL) {
				perror("[line_splitter] malloc");
				exit(1);
			}
			strlcpy(commandargs[argindex], tmp, size+1);
		} else {
			break;
		}

		argindex++;
		if (argindex >= argmax)
			break;
	}
	commandargs[argindex] = NULL;
	return argindex;
}

void dump_state(void)
{
	ylog(VERB_FORCE,
	"dump_state() requested, dump follows\n"
	"\n-- GENERAL\n"
	"s = %d\n"
	"configfile = %s\n"
	"opt_udp_port = %d\n"
	"local_size = %u\n"
	"local_used = %u\n"
	"\n-- FORWARDING\n"
	"opt_forward = %d\n"
	"dns_forward_port = %d\n"
	"forward_server = %s\n"
	"forward_count = %d\n"
	"forward_size = %u\n"
	"forward_used = %u\n"
	"\n-- CACHE\n"
	"opt_cache = %d\n"
	"cache_count = %d\n"
	"cache_table_size = %u\n"
	"cache_table_used = %u\n"
	"\n-- STATS\n"
	"statistic_received_packet = %d\n"
	"statistic_invalid_packet = %d\n"
	"statistic_query_count = %d\n"
	"statistic_iquery_count = %d\n"
	"statistic_response_count = %d\n"
	"\n",	s, configfile, opt_udp_port, local_table.size, local_table.used,
		opt_forward, dns_forward_port, inet_ntoa(forward_server[0]),
		forward_count, forward_table.size, forward_table.used,
		opt_cache, cache_count, cache_table.size, cache_table.used,
		statistic_received_packet, statistic_invalid_packet,
		statistic_query_count, statistic_iquery_count,
		statistic_response_count);
	fflush(logfp);
}

/* Don't expect maximun security here, anyway the id is 16 bit large */
u_int16_t get_rand_id(void)
{
	u_int32_t id;
	static u_int16_t inc = 0;
#ifdef __unix__
	struct timeval tmptv;

	gettimeofday(&tmptv, NULL);
	id = tmptv.tv_usec ^ tmptv.tv_sec ^ getpid() ^ inc++;
#else /* not __unix__ */
	id = rand() ^ time(NULL) ^ clock() ^ inc++;
#endif /* not __unix__ */
	return (u_int16_t) (id & 0xffff);
}

/* WARNING: THIS MAY WORK ONLY WITH GNU MALLOC
   but this stuff is used only to trace memory leaks
   and isn't part of a normal ENS binary */
#ifdef TRACE_LEAKS
#undef malloc
#undef realloc
#undef free
void tl_current(int x, int y);

void *tl_malloc(char *file, int line, size_t size)
{
	void *ptr = malloc(size);
	int *l = (int*) ptr;
	printf("%s %d: malloc(%d) = %p, ", file, line, size, ptr);
	if (size == 0) {
		printf("malloc zero\n");
		exit(1);
	}
	printf("allocated %d bytes\n", *(l-1));
	tl_current(*(l-1), 1);
	return ptr;
}

void *tl_realloc(char *file, int line, void *ptr, size_t size)
{
	int *o = (int*) ptr;
	int old = (o != NULL) ? *(o-1) : 0;
	void *newptr = realloc(ptr, size);
	int *l = (int*) newptr;
	printf("%s %d: realloc(%p, %d) = %p, ", file, line, newptr, size, ptr);
	printf("allocated %d bytes\n", *(l-1) - old);
	tl_current(*(l-1)-old, ptr ? 0 : 1);
	return newptr;
}

void tl_free(char *file, int line, void *ptr)
{
	int *l = (int*) ptr;
	printf("%s %d: free(%p), ", file, line, ptr);
	printf("freed %d bytes\n", *(l-1));
	tl_current(-(*(l-1)), -1);
	free(ptr);
}

void tl_current(int x, int y)
{
	static int current = 0;
	static int chunkes = 0;

	current += x;
	chunkes += y;
	printf("current: %d bytes (in %d chunkes)\n", current, chunkes);
}
#endif
